packages/dds/tree/src/feature-libraries/flex-map-tree/mapTreeNode.ts

```diff
@@ -0,0 +1,744 @@
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/core-utils/internal";
+import {
+	AnchorNode,
+	EmptyKey,
+	FieldKey,
+	MapTree,
+	TreeNodeSchemaIdentifier,
+	TreeValue,
+	Value,
+} from "../../core/index.js";
+import { brand, fail, mapIterable } from "../../util/index.js";
+import {
+	FlexTreeContext,
+	FlexTreeEntityKind,
+	FlexTreeField,
+	FlexTreeFieldNode,
+	FlexTreeLeafNode,
+	FlexTreeMapNode,
+	FlexTreeNode,
+	FlexTreeNodeEvents,
+	FlexTreeOptionalField,
+	FlexTreeRequiredField,
+	FlexTreeSequenceField,
+	FlexTreeTypedField,
+	FlexTreeTypedNode,
+	FlexTreeTypedNodeUnion,
+	FlexTreeUnboxField,
+	FlexTreeUnboxNodeUnion,
+	FlexibleFieldContent,
+	FlexibleNodeSubSequence,
+	TreeStatus,
+	flexTreeMarker,
+	indexForAt,
+} from "../flex-tree/index.js";
+import {
+	FlexAllowedTypes,
+	FlexFieldNodeSchema,
+	FlexFieldSchema,
+	FlexMapNodeSchema,
+	FlexTreeNodeSchema,
+	LeafNodeSchema,
+	isLazy,
+	schemaIsFieldNode,
+	schemaIsLeaf,
+	schemaIsMap,
+	schemaIsObjectNode,
+} from "../typed-schema/index.js";
+import { FlexImplicitAllowedTypes, normalizeAllowedTypes } from "../schemaBuilderBase.js";
+import { FlexFieldKind } from "../modular-schema/index.js";
+import { FieldKinds, SequenceFieldEditBuilder } from "../default-schema/index.js";
+import { createEmitter } from "../../events/index.js";
+
+// #region Nodes
+
+/** A node's parent field and its index in that field */
+interface LocationInField {
+	readonly parent: MapTreeField<FlexAllowedTypes>;
+	readonly index: number;
+}
+
+/**
+ * A readonly implementation of {@link FlexTreeNode} which wraps a {@link MapTree}.
+ * @remarks Any methods that would mutate the node will fail,
+ * as will the querying of data specific to the {@link LazyTreeNode} implementation (e.g. {@link FlexTreeNode.context}).
+ * MapTreeNodes are unconditionally cached -
+ * when retrieved via {@link getOrCreateMapTreeNode}, the same {@link MapTree} object will always produce the same `MapTreeNode` object.
+ */
+export class MapTreeNode<TSchema extends FlexTreeNodeSchema> implements FlexTreeNode {
+	public readonly [flexTreeMarker] = FlexTreeEntityKind.Node as const;
+
+	private readonly events = createEmitter<FlexTreeNodeEvents>();
+
+	/**
+	 * Create a new MapTreeNode.
+	 * @param schema - the schema of this node
+	 * @param mapTree - the {@link MapTree} containing the data that this node wraps
+	 * @param location - the parentage of this node, if it is being created underneath an existing node and field, or undefined if not
+	 * @remarks This class (and its subclasses) should not be directly constructed outside of this module.
+	 * Instead, use {@link getOrCreateMapTreeNode} to create a MapTreeNode from a {@link MapTree}.
+	 */
+	public constructor(
+		public readonly schema: TSchema,
+		public readonly mapTree: MapTree,
+		private location: LocationInField | undefined,
+	) {
+		assert(!nodeCache.has(mapTree), "A node already exists for the given MapTree");
```

Comment:
Seems like the constructor should probably call out this invariant / behavior.

Reply:
Agreed, done.